daemon/graphdriver/devmapper/deviceset.go

```diff
@@ -841,11 +841,62 @@ func (devices *DeviceSet) createRegisterDevice(hash string) (*devInfo, error) {
 	return info, nil
 }
 
-func (devices *DeviceSet) createRegisterSnapDevice(hash string, baseInfo *devInfo, size uint64) error {
+func (devices *DeviceSet) takeSnapshot(hash string, baseInfo *devInfo, size uint64) error {
+	var doSuspend bool
+	deferredRemove := true
+
 	if err := devices.poolHasFreeSpace(); err != nil {
 		return err
 	}
 
+	if !devices.deferredRemove {
+		deferredRemove = false
+	}
+
+	devinfo, err := devicemapper.GetInfoWithDeferred(baseInfo.Name())
+	if err != nil {
+		return err
+	}
+
+	if devinfo != nil && devinfo.DeferredRemove == 0 {
+		deferredRemove = false
+	}
+
+	if deferredRemove {
+		err = devices._cancelDeferredRemoval(baseInfo)
+		if err != nil {
+			if err != devicemapper.ErrEnxio {
+				return err
+			}
+			// Error is ErrEnxio. Device is probably already gone. Return success.
+			deferredRemove = false
+		}
+	}
+
+	// if deferred removal was cancelled, we can safely assume the device is active.
+	if deferredRemove {
+		doSuspend = true
+		defer devices.deactivateDevice(baseInfo)
+	} else {
+		devinfo, _ := devicemapper.GetInfo(baseInfo.Name())
+		doSuspend = devinfo != nil && devinfo.Exists != 0
+	}
```

Comment:
We don't have to do GetInfo() again. We already did it once and that will give information if device exists or not.